pghoard/common.py

```diff
@@ -100,6 +104,55 @@ class BaseBackupMode(StrEnum):
     pipe = "pipe"
 
 
+class ProgressData(BaseModel):
+    current_progress: float = 0
+    last_updated_time: float = 0
+
+    @property
+    def age(self) -> float:
+        return time.time() - self.last_updated_time
+
+    def update(self, current_progress: float) -> None:
+        self.current_progress = current_progress
+        self.last_updated_time = time.time()
+
+
+class PersistedProgress(BaseModel):
+    progress: Dict[str, ProgressData] = Field(default_factory=dict)
+    _lock: threading.Lock = threading.Lock()
+
+    @classmethod
+    def read(cls, metrics: Metrics) -> "PersistedProgress":
+        if os.path.exists(PROGRESS_FILE):
+            with open(PROGRESS_FILE, "r") as file:
+                try:
+                    return cls.parse_raw(file.read())
+                except Exception as ex:  # pylint: disable=broad-except
+                    LOG.exception("Failed to read persisted progress file: %r", ex)
+                    metrics.unexpected_exception(ex, where="read_persisted_progress")
+        return cls()
+
+    def write(self, metrics: Metrics):
+        with self._lock:
+            try:
+                with tempfile.NamedTemporaryFile("w", delete=False, dir=os.path.dirname(PROGRESS_FILE)) as temp_file:
+                    temp_file.write(self.json())
+                    temp_path = temp_file.name
+                os.rename(temp_path, PROGRESS_FILE)
+            except Exception as ex:  # pylint: disable=broad-except
+                LOG.exception("Failed to write persisted progress file: %r", ex)
+                metrics.unexpected_exception(ex, where="write_persisted_progress")
+
+    def get(self, key: str) -> ProgressData:
+        self.progress.setdefault(key, ProgressData())
+        return self.progress[key]
+
+    def reset(self, key: str, metrics: Metrics) -> None:
+        if key in self.progress and self.progress[key] != ProgressData():
```

Comment:
why do check against and the default value and not delete it? 
Is it possible for there to be backup files of zero bytes?

Reply:
zero-byte backup files should not exist, will remove the check